Feature: Add a new Card

sviat-barbutsa opened the pull request.

sviat-barbutsa committed:
- Card model
- Modal component
- AddNewCardForm with the dialog
- common form elements styling
- common buttons styling
- implemented initial Card instantiating at the Modal open

sviat-barbutsa committed:
- add Card functionality
- Card item render
- buttons global UI common styles

nebosite commented on mvvm_cards/src/models/AppModel.ts:
```diff
@@ -21,7 +22,18 @@ import { IAppModel } from "./i_appmodel";
     get docTitle() { return this._docTitle; }
     set docTitle(title: string) { this._docTitle = title; }
+
+    @observable
```
Nice abstraction for the cards.  

Getter and setter won't be necessary as this is an observable collection and the interactions will be 100% add/remove.

sviat-barbutsa replied:
Removed

nebosite commented on mvvm_cards/src/models/AppModel.ts:
```diff
@@ -21,7 +22,18 @@ import { IAppModel } from "./i_appmodel";
+     // in the current Mobx version we might not to write the action. Result will be the same
+    @action
+    addCard(card: ICard) { 
+        this._cards.unshift(card);
```
curious:  why unshift instead of push?

sviat-barbutsa replied:
it's because of our requirements. We need to add a new card at the top of a column

so let's think about a workflow
we'll have an array of cards and we'll iterate it in the View. So the item with the index 0 will be at the top (first item) and the item with the last index will be at the bottom.
So when we'll add a new item, the View will be re-rendered and the iteration cycle will happen again. In this case if we push our new item like the last item  ( and push adds the item ato the end of the array ) so in the View, it also will be rendered at the bottom of the column.

so that's why we need to at the new item to the beginning of our cards array (to render it at the top)

![image](https://user-images.githubusercontent.com/9897782/72661142-8e882a00-39df-11ea-8f08-b06f984251e4.png)

nebosite commented on mvvm_cards/src/shared/Modal.tsx:
```diff
@@ -0,0 +1,29 @@
+import * as React from "react";
```
Let's rename this file to be more clear.  Maybe "ModalDialog.ts"   Could the dialog be both Model and Modeless?  Just different in the way it is invoked?

sviat-barbutsa replied:
To be honest the naming of such Components alway was some funny things for me. Some people say that everything that appears like a Popup is a Model, some people say that if the Popup has some Actions that user must execute (eg. Save & Exit) and if the background is blocking - so it's a Dialog.

I prefer to follow next:
1) a Modal is a Component that appears outside of the document flow (like an independent component). It could be some Additional Info, some text, image, etc.
2) a Dialog is a Subset of the Modal. It should also appear outside of the document but in addition to the Modal it should have some actions that user should execute to do something with the dialog and the document background should be disabled

So I called it Dialog first, but as you wrote that you want to see Modal in the task description, so later I decided to rename Dialog to Modal =) just wanted to make everything exactly like you wanted

Dialog can't be modeless because it should appear outside of the document.
In general, it isn't a problem at all, if you want to rename it to ModalDialog, I'll do it. It's a very very minor thing.

https://material-ui.com/components/dialogs/

![image](https://user-images.githubusercontent.com/9897782/72661321-1e7aa380-39e1-11ea-9bdf-3fc2d98704f2.png)

sviat-barbutsa replied:
Renamed it to ModalDialog

nebosite replied:
FYI - In WPF and other windows frameworks, a dialog is a little child window that floats above the other UI.   Dialogs can be either modal or modeless.  Modal means that the UI behind the dialog is paused while the dialog is open.  Modeless means that the dialog UI operates in parallel with the main application UI.   So a save dialog would typically be modal and something like a tool selector dialog would typically be modeless.

Specifically, there is no concept of a window class that is specifically one or the other.  The programmer chooses to open any window as modal or modeless, depending on the need.   Not sure if createPortal supports this concept.   Also not sure the best way to create something modeless. Maybe window.open?

nebosite commented on mvvm_cards/src/shared/Modal.tsx:
```diff
@@ -0,0 +1,29 @@
+export default function Modal(props: Props) {
+  const { children, onDecline, onConfirm } = props;
+
+  const bodyEl = document.querySelector("body");
```
nit: Please give this field a long name

sviat-barbutsa replied:
Sure, done

nebosite commented on mvvm_cards/src/shared/Modal.tsx:
```diff
@@ -0,0 +1,29 @@
+  const bodyEl = document.querySelector("body");
+
+  return (
+    ReactDOM.createPortal(
```
Can the modality be controlled here?

sviat-barbutsa replied:
in general, it should be like you described but I thought that in this case, it will be redundant because we need to pass the _open_ prop through few parent components. So that's why I made it so simple.

if you want me to do it so I can also move this 'showDialog' property to the AppModel and in this case, we won't need to pass this property through all parent components.

But, to be honest, I'd like rather create a separated CardsStore (eg. CardsModel) where we could keep al info regarding our particular CardModel

I don't like the solution to push everything to the one AppModel. I described the reason before

So what case would you like to see?

nebosite replied:
Since the code is keeping the dialog around, I would like the visibility of the dialog controlled by wether or not it is holding on to an ICard object.  So ```dialog.card = null``` would automatially hide the dialog and setting to to a card would automatically show it.

sviat-barbutsa replied:
It could be done for sure, but in this case, this Dialog won't be re-usable because of the strict Card content dependency. 
The common use case for re-usable Dialog is just using the *show*  boolean property and show the dialog regarding this prop value.
eg.

```
      <ModalDialog
           show={isShow}
       ....
      >
```

But if you want to show/hide this dialog based on the content (if the content presented) so I can do it as well. I'd like to do it generically, not for cards only
We can do it via _children_ property. If it exists - we'll show the dialog, if not - hide. In this case, we can use Dialog not only for Card but for any other content type as well
Anyway, it will add additional complexity

I'll implement this content-based control and if you won't like it I'll change it to the Card only (I'll do it in my free time)

sviat-barbutsa replied:
Done, but I'm not sure that it's exactly what you wanted, so, could you please check it and if it isn't that you expected, I'll change it in my free time.
Btw. I don't much like these changes that I did because it's a bit redundant (just my opinion)

Maybe you wanted to add the state Show/Hide directly to the ModalDialog and based on the children property (or Card) it should change its show property to false/true?

nebosite commented on mvvm_cards/src/views/Card.tsx:
```diff
@@ -0,0 +1,12 @@
+import * as React from "react";
+import { ICard } from "models/i_card";
+
+export default function Card(card: ICard) {
```
Let's call this a CardView or maybe a CardMiniView

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+import * as React from "react";
```
Let's name this "NewCardDialog"

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+  hide: () => void;
+}
+
+export function AddNewCardModal(props: Props) {
```
Would like to have these be classes if possible.  What are the pros and cons of that from your perspective?

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+const { useState, useRef } = React;
+
+type Props = {
+  appModel?: IAppModel;
```
rather than passing in the appModel, pass in a newly created ICard object so we can bind directly to that.

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+
+export function AddNewCardModal(props: Props) {
+  const { hide, appModel } = props;
+  const cardEntity = new Card();
```
This dialog should only know how to manipulate an ICard.   The AppModel is the sole factory for creating new cards.

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+  const cardEntity = new Card();
+  // a reference to our form that we'll submit manually from our modal
+  const formRef = useRef(null);
+  const [ card, setCard ] = useState(cardEntity);
```
Probably not necessary if the card is being passed in?

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+  const [ card, setCard ] = useState(cardEntity);
+
+
+  const generateOnChange = (fieldName: string) => (e: React.ChangeEvent<HTMLInputElement>) => {
```
This is very clever and I like it.     However, for this particular application, I'd like to be more conventional and have explicit functions "onNameChanged", "onDescriptionChanged"  etc.  It will make the structure more clear for people unfamiliar with the code.

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+
+  return (
+    <Modal 
+      onDecline={hide} 
```
To have a more windows-like pattern, I think I'd like to have a single event, onClosed, that receives an enum with the state of the closed dialog- either "Save" or "Decline"

nebosite commented on mvvm_cards/src/views/toolbar/AddNewCardModal.tsx:
```diff
@@ -0,0 +1,81 @@
+          // in general we could avoing the form submitting at all 
+          // and do it in the modal onConfirm method
+          // but I just wanted to show you how can you handle forms
+          appModel.addCard(card);
```
Move this code into the caller based on the return status of this dialog.  The view that spawned this dialog gets to decide what to do with the card.    The general motivation for this is that the new card dialog could be used anywhere, so it should not "know" that there is an appmodel collecting cards.  For example, I could have a little side control that also collects cards, but not as part of the main colleciton.

sviat-barbutsa replied:
Sure. Done
(Won't track time for it and next things as I promised)

nebosite commented on mvvm_cards/src/views/toolbar/MainDocumentPageToolbar.tsx:
```diff
@@ -1,18 +1,26 @@
 
+const { useState, useCallback } = React;
 
 function MainDocumentPageToolbar(props: {appModel?: IAppModel}) {
```
Again, would like to make these classes, but curious to know why I might want functions over a class.  
Main reason here is that my audience is going to be people who think primarily in classes and I want to code forms to not be a stumbling block to understanding.

sviat-barbutsa replied:
I thought you want to learn about Hooks
Anyway, reworked it to classes

( Didn't track time for it as I promised )

nebosite replied:
I'm still digging into hooks and figuring out how is plays with MVVM concepts of clean separation of logic.    I'm also trying to understand how to make use of inheritance concepts with functional components.

nebosite commented on mvvm_cards/src/assets/styles/index.css:
```diff
@@ -8,6 +8,19 @@ html, body {
   display: flex;
 }
 
+.btn {
```
nit:  Please give this a long name

sviat-barbutsa replied:
Done

nebosite commented on mvvm_cards/src/views/toolbar/MainDocumentPageToolbar.tsx:
```diff
@@ -32,10 +40,11 @@ function MainDocumentPageToolbar(props: {appModel?: IAppModel}) {
+      <button className="btn" onClick={() => alert("The Save functionality is coming soon")}>Save</button>
+      <button className="btn" onClick={() => alert("The Load functionality is coming soon")}>Load</button>
+      <button className="btn" onClick={() => alert("The New functionality is coming soon")}>New</button>
+      { showModal && <AddNewCardModal hide={hide}/> }
```
What are the pros and cons of generating a new dialog each time instead of showing and hiding a single dialog?

sviat-barbutsa replied:
It's a very good point.  To be honest, I also don't like this solution but other solutions will look a bit more ugly.

I did it because in our dialog we are creating a new Card and passing it to the state of the component at the creating stage of the component.
After that, we don't destroy this component so the state continues to live with your previous Card even when we hide the Dialog. It wasn't been removed, only hidden so the state is going to be the same.
To solve this issue we can add some conditional logic to our component that will track 
if ->  show dialog -> new Card -> put it to the state
and
if -> hide dialog -> reset the state

also, it could be done with some useEffect callback etc.

so I decided that this solution will be a bit complicated than just mount & unmount a modal dialog. About performance, it's a bit worse than just manipulation via Show\Hide because DOM operations are expensive, but in this particular case it's simpler and it's about Sanity.
So I thought you will like this solution.

If you want me to re-work it the show/hide instead mount/unmount - I'll do it, it isn't a problem
So what do you think?

nebosite replied:
Leave it for now.  I will play with this later.

sviat-barbutsa replied:
As we changed our component to ClassBased I was able to add a lifecycle hook to control the state behavior. It added additional logic but in this case, it's possible to remove mount/unmount render logic and use show/hide with the correct state.

nebosite commented on mvvm_cards/src/assets/styles/index.css:
```diff
@@ -68,3 +87,86 @@ html, body {
+  margin-top: 30px;
+}
+
+.modal-decline-btn {
```
expand btn

sviat-barbutsa replied:
done, renamed it to _button_

sviat-barbutsa reviewed (commented):
Fixed all the things that you mentioned and about 2 of them I need to know about your preferences. After you give your answer I'll implement the chosen solution.

sviat-barbutsa committed:
- reworked function to classes
- changed short naming to explicit
- removed model.addCard functionality from the AddCardDialog to the parent

nebosite reviewed (approved):
Thansk for the changes!  Let me know when this PR is ready to merge.

sviat-barbutsa committed:
- added to a ModalDialog the children based contidion to create a portal or not
- reworked the mount/unmout logic of the addNewCardModalDialog showing it just only show/hide condition. Added a lifecycle hook to controls the state with the new Card based on the show/hide property

sviat-barbutsa commented:
uhm, this PR is still active ) 
We have 1 unresolved conversation here, but, as I understand, everything fine now.
I'll move all changes from this PR to our D-N-D branch to keep them synced
